### DISCUSSION TITLE
Fallback link on desktop

### DISCUSSION THREAD

**Discussion by phillipcmittmann** [invertase/react-native-firebase#5737] [Q&A] [answered]
So, I'm trying to figure out how to configure a fallback link, when I manually create a dynamic link, that redirects in desktops. When using fallbackUrl in the builder, the Promise returned contains both AFL and IFL parameters not encoded.

```
const link = await dynamicLinks().buildLink({
                link: `https://myapp.example.com/todayEdition/${Moment().format('YYYYMMDD')}`,
                domainUriPrefix: 'https://myapp.example.com',
                android: {
                    packageName: 'com.myapp.example,
                    fallbackUrl: 'www.google.com.br'
                },
                ios: {
                    bundleId: 'com.myapp.example',
                    fallbackUrl: 'www.google.com.br'
                },
            });
```

The snipped above results in:

```
https://myapp.example.com?afl=www.google.com.br&apn=com.myapp.example&ibi=com.myapp.example&ifl=www.google.com.br&link=https%3A%2F%2Fmyapp.example.com%2FtodayEdition%2F20210923
```

When opening in mobile, the links redirects/open the application like it should. On desktop, it says: "Dynamic Link Not Found
Failed to resolve uri domain prefix: https://myapp.example.com/todayEdition", instead of redirecting to Google as expected.

The domain prefix is already registered in the console. 

React-Native - 0.64.2
Dynamic Links - 12.1.0

**Comment by mikehardy** [comment]
Hi there!
Interesting - first thing I can say is in general approaching support in any fashion with outdated versions is always going to result in an immediate request to update and make sure you still reproduce. 12.8.0 is current

I don't want to spend any more time on it really until I know it's still valid in current versions. Past that - I think the assumption is that you need to url-encode all links that you send in to the builder? Does it work if you do that?

**Comment by phillipcmittmann** [comment]
First, sorry about the issue. Couldn't figure out how to delete a discussion because I created by mistake and didn't know if this was the right place to have this conversation.

>When opening in mobile, the links redirects/open the application like it should.

Again, sorry. I meant redirects to the store* like it should.


After updating the dependency, all the parameters passed to the build are encoded.
```
https://myapp.example.com/?ibi=com%2Eexample%2Emyapp&apn=com%2Eexample%2Emyapp&afl=www%2Egoogle%2Ecom%2Ebr&ifl=www%2Egoogle%2Ecom%2Ebr&link=https%3A%2F%myapp%example%2Ecom%todayEdition%2F20210923
```

**Reply by mikehardy** [reply]
Oh great - so, this appears resolved via current version? Just want to make sure before marking it as answer

**Reply by phillipcmittmann** [reply]
Not yet. On desktops still Dynamic Link Not Found. The redirect was already working as intended for mobiles (redirecting for the store as expected). Now, with 12.8.0, it stopped working.

EDIT: the link works if I already have the app installed. Otherwise, it doesn't open the browser on android, and loads into Dynamic Link Not Found on iOS.

**Reply by mikehardy** [reply]
Hmm - have you tried putting the link into a browser URL with `?d=1` on it? https://firebase.google.com/docs/dynamic-links/debug
I could have sworn I tested all this for my own work project when I integrated it, but it appears from 12.1.0 to 12.8.0 perhaps there were behavior changes, and/or perhaps something is wrong here, I'm just not sure sorry. May be that your opening an issue was actually the correct course as this could be a bug

**Reply by phillipcmittmann** [reply]
In `https://myapp.example.com/todayEdition?d=1` shows the debug as expected. Adding a parameter (`/todayEdition/20210923?d=1`) results in Dynamic Link not found, as it should because this link isn't registered on the console, just `/todayEdition`. Still, wont redirect to the store on mobile like before. Removing the OS specific parameters didn't made a difference.

**Comment by phillipcmittmann** [comment, marked as answer]
After a bit more of testing I got to redirect on desktops after manually adding `&ofl=ENCODED_FALLBACK_LINK` at the end of new dynamic link.
I'm not sure how the links are encoded in the builder even after looking at the source code. It's a bit of work, but I can manually encode the links I need in the app, verifying which one I need per occasion.
On that note, in my opinion, it should be created a optional parameter `desktop` at buildLink like `android` or `ios` to act as fallback link for desktops.

EDIT: formating and semantics
EDIT 2: `encodeURIComponent` doesn't encode dots, but resolved the problem anyway

**Reply by mikehardy** [reply]
Interesting! I see this was added quite some time ago:
- firebase-ios-sdk 4.7.0 in 2017 added it https://firebase.google.com/support/release-notes/ios#dynamic-links_36

But then I look in the iOS SDK reference and cannot find it there or in the code, and I can't find it in the Android one either.
There is no dynamic links module for firebase-js-sdk (which would normally give us a pattern to work off of, if we make an API that isn't represented by the native iOS/Android SDKs and just implement it ourselves)

So, I'm not sure how we can support this well but I'll create an enhancement issue at least to capture this

**Reply by Nehal-Sanklecha** [reply]
@phillipcmittmann  Could you share an example of how you've manually added the ofl in the link?

```
Const url = await dynamicLinks().buildShortLink({
        android: {packageName: getBundleId()},
        ios: {bundleId: getBundleId(), appStoreId: ENV.APP_STORE_ID},

        domainUriPrefix: ENV.DYNAMIC_LINK_URI_PREFIX,
        link: encodeURI(`https://www.example.com/app?id=${id}&name=${name}&ofl=${fallbackUrl}`),
    }, dynamicLinks.ShortLinkType.SHORT);

```
But this is not working. Could you help me here?

**Reply by cristiam86** [reply]
@mikehardy could you please confirm that this only works on the long version of the dynamic link and there is no way of having a short link with the `ofl` parameter?

**Reply by mikehardy** [reply]
Unfortunately @cristiam86 I cannot - I *think* the relationship between long links and short links is that the short link is simply an encoded version of anything you put in the long link, so I *think* that ofl - if it is valid in a long link - will be in the short link if you put it there but the word "confirm" implies to me that I have personally tested it and know that it is working. I haven't tested my dynamic links implementation in my work projects in so long that I will not confirm it authoritatively.

If I had the need to confirm it as it appears you do, I would test it exhaustively, check the relevant code in firebase-ios-sdk / firebase-android-sdk (assuming these parts are open...I think they are?) and/or ask on stackoverflow with the appropriate tags so their support will respond

I recognize this is essentially a non-answer, but saying you don't know when you don't know is important and there's a path forward here for you

**Reply by cristiam86** [reply]
Thank you @mikehardy.

I've been testing it more, and I believe we had it wrong since the beginning (maybe it is not clear enough in the docs).
We have a mobile app without a web version and were assuming that when calling `buildLink` and `buildShortLink` we should have the same domain for `link` and `domainUriPrefix` (default Firebase Dynamic Links prefix <app>.page.link).

Now that I've also seen your answer [here](https://github.com/invertase/react-native-firebase/discussions/4479#discussioncomment-118164), it seems that the link pointing to the web domain is the right thing to do because if you have informed the `ios` and `android` params when creating the link, they are going to open the app or redirect the user to the store in case it is not installed.

**Reply by mikehardy** [reply]
@cristiam86 that sounds about right, and it was amusing (to me) reading my own answer from 2020, I have forgotten most of what I've written here, the volume is too high to remember it 😆 . I'm glad it helped, and now that I am familiar (again!) with the implementation in my own app (which I had also forgotten..."it just works") I think your analysis is correct, just point to the web domain with the "link"

**Reply by harrysayers** [reply]
Hey @cristiam86,

Thanks for the pointers, I'm trying what you've suggested however on desktop I'm still not being redirected to my website but instead receiving a 'invalid dynamic link'. 

Below is my function that takes in some params and returns a link - What am I doing wrong?

If I replace 'link' with ```https://applandingpage/${suffix}/${params}``` the link returns a 404 because the page doesn't exists on my landing page site and it stops working for the app.

Any help is appreciated :) 

```
async function buildLink(suffix, params, campaign, socialTitle, socialImage, descriptionText) {

        const link = await dynamicLinks().buildShortLink({
          link: `https://app.page.link/${suffix}/${params}',
          domainUriPrefix: `https://app.page.link`,
          social:{
             title: socialTitle,
             imageUrl: socialImage,
             descriptionText: descriptionText
          },
          iOS:{
            bundleId: '12345',
            fallbackUrl: 'https://applandingpage.com',
          },
          android: {
            packageName: 'app',
            fallbackUrl: 'https://applandingpage.com',
          },
          analytics: {
            campaign: campaign,
          },
        });
      
        return link;
}
```

**Reply by harrysayers** [reply]
@mikehardy @phillipcmittmann any help on the above question, would be greatly appreciated! :)

**Reply by cristiam86** [reply]
hello @harrysayers the URL on the `link` property should point to your website URL whatever it is. Right now you are using the same URL prefix that you currently use for generating Dynamic Links on firebase.

**Reply by harrysayers** [reply]
@cristiam86 Thank you for the reply. 

I've used my website as the link property and the domainUriPrefix handles the deep link as below: 

```
link: `https://www.mywebsite.com`,
domainUriPrefix: `https://myapp.page.link/${suffix}/${params}`
```
However I'm getting a "The operation couldn’t be completed. Long link is not parsable" error returned when I'm trying to create a link... so what I'm doing above can't be correct.

**Reply by cristiam86** [reply]
Can you please paste the actual code? Also, I'm not using the `fallbackURL` property since it doesn't redirect users to the Play/Apple Stores.
Also, please note that on iOS, you are missing the `appStoreId` property.

**Reply by harrysayers** [reply]
@cristiam86 

Yes, of course. This is the function which I pass some params to and it returns a link, I've currently got the link and domainUriPrefix setup as you suggested.

I've replaced the actual website link with www.mywebsite.com and the app deep link with myapp.page.link.


Yes, good spot about the appStoreId... I've added that in, thank you.

```
async function buildLink(suffix, params, campaign, socialTitle, socialImage, descriptionText) {

  const link = await dynamicLinks().buildShortLink({
    link: 'https://www.mywebsite.com',
    domainUriPrefix: 'https://myapp.page.link/${suffix}/${params}',
    social:{
       title: socialTitle,
       imageUrl: socialImage,
       descriptionText: descriptionText
    },
    ios:{
      bundleId: '12345',
      appStoreId: '12345',
      fallbackUrl: 'https://mywebsite.com',
    },
    android: {
      packageName: 'myappname',
      fallbackUrl: 'https://mywebsite.com',
    },
    analytics: {
      campaign: campaign,
    },
  });

  return link;

}
```

Thanks for any help. :)

**Reply by cristiam86** [reply]
@harrysayers do you own myapp.page.link as URL prefix in Firebase? Otherwise, you need to create one in the web console and use it here.

**Reply by harrysayers** [reply]
@cristiam86 yeah I own myapp.page.link ( its registered in firebase) and it works when its set as the link parameter however when moved to the DomainUriPrefix it does not

**Comment by cristiam86** [comment]
@harrysayers `domainUriPrefix` should be only that prefix. The ${suffix}/${params} should be included in the `link` property

**Reply by harrysayers** [reply]
@cristiam86 I've tried this but I'm just receiving a 404 on my website when the link is clicked because "https://www.mywebsite.com/suffix/params doesn't exist and the app doesn't open the correct page because it's not receiving the suffix and params. I do apologies, but something isn't right here.

``` 
link: `https://www.mywebsite.com/${suffix}/${params}`,
domainUriPrefix: `https://myapp.page.link`,
```

**Reply by cristiam86** [reply]
https://www.mywebsite.com/suffix/params should exist on your website.
The app should also handle this URL to know to which screen send the user to.
I'm sorry I'm not able to help you better. Maybe explaining the whole app can help, and also provide any code example or prototype.
For me it works as expected :)

**Reply by harrysayers** [reply]
@cristiam86 thanks for your help. 

I got it working, I forgot to add my https://www.mywebsite.com to my app.js linking prefixes, it now opens the correct screen with params in app like it used too. My website still doesn't handle suffix/params correctly and returns a 404, I don't know what's going on there.... as a hacky fix I've just redirected all 404's to the homepage (my website is just a landing page, so it doesn't do anything fancy with content or require the params).

Hope this helps someone, maybe :)
